Marlin/planner.cpp

```diff
@@ -1196,6 +1218,47 @@ void Planner::_buffer_steps(const int32_t (&target)[ABCE], float fr_mm_s, const
     // Start with print or travel acceleration
     accel = CEIL((esteps ? acceleration : travel_acceleration) * steps_per_mm);
 
+    #if ENABLED(LIN_ADVANCE)
+      /**
+       *
+       * Use LIN_ADVANCE for blocks if all these are true:
+       *
+       * esteps             : This is a print move, because we checked for A, B, C steps before.
+       *
+       * extruder_advance_K : There is an advance factor set.
+       *
+       * de > 0             : Extruder is running forward (e.g., for "Wipe while retracting" (Slic3r) or "Combing" (Cura) moves)
+       */
+      block->use_advance_lead =  esteps
+                              && extruder_advance_K
+                              && de > 0;
+
+      if (block->use_advance_lead) {
+        block->e_D_ratio = (target_float[E_AXIS] - position_float[E_AXIS]) /
+          #if IS_KINEMATIC
+            block->millimeters
+          #else
+            SQRT(sq(target_float[X_AXIS] - position_float[X_AXIS])
+               + sq(target_float[Y_AXIS] - position_float[Y_AXIS])
+               + sq(target_float[Z_AXIS] - position_float[Z_AXIS]))
```

Comment:
Since `block->millimeters` is now being provided by the `millimeters` argument to `_buffer_steps` this extra calculation may no longer be needed. See above where it does this…
```cpp
  if (block->steps[A_AXIS] < MIN_STEPS_PER_SEGMENT && block->steps[B_AXIS] < MIN_STEPS_PER_SEGMENT && block->steps[C_AXIS] < MIN_STEPS_PER_SEGMENT) {
    block->millimeters = FABS(delta_mm[E_AXIS]);
  }
  else if (!millimeters) {
    block->millimeters = SQRT(
      #if CORE_IS_XY
        sq(delta_mm[X_HEAD]) + sq(delta_mm[Y_HEAD]) + sq(delta_mm[Z_AXIS])
      #elif CORE_IS_XZ
        sq(delta_mm[X_HEAD]) + sq(delta_mm[Y_AXIS]) + sq(delta_mm[Z_HEAD])
      #elif CORE_IS_YZ
        sq(delta_mm[X_AXIS]) + sq(delta_mm[Y_HEAD]) + sq(delta_mm[Z_HEAD])
      #else
        sq(delta_mm[X_AXIS]) + sq(delta_mm[Y_AXIS]) + sq(delta_mm[Z_AXIS])
      #endif
    );
  }
  else
    block->millimeters = millimeters;
```

Reply:
Following from that, you may no longer need `target_float` or `position_float` either.

Reply:
It looks like the XYZ parts of `target_float` and `position_float` are almost certainly not needed.

Reply:
Be careful, the `millimeters` argument is only given by the delta code as far as I remeber the PR. usual print moves will not contain it.
I also don't think it makes sense to precalculate it for any possible case as we have to calculate it once anyway. And I think the right context for this is the planner..

Reply:
Ah, right. So, is this calculated value also not useful?
```cpp
SQRT(sq(delta_mm[X_AXIS]) + sq(delta_mm[Y_AXIS]) + sq(delta_mm[Z_AXIS]))
```

Reply:
Using the step values instead of float (and this is not much more, as it's steps multiplied by mm/step) made a lot of problems in the past, therefore I dislike it. I know it could save a lot of cycles but it might cost a lot of time tracking down stange behaviours when LIN_ADVANCE is used.
Small example. All we want to know is the ratio between print move length D and extruded filament length E. Let's say we have a printer with 100steps/mm on X and Y and 400 steps/mm on E. Let's start from all axis = 0:

G1 X0.05 Y0.1 E0.00148 // a typical 3mm filament move with 10 steps inside the move, so well above the limit

This equals a ratio E/D of 0.01324.

Now we convert to steps:
X = 5
Y = 10
E = 1 (instead of 0.72)

Now we convert back to mm again:
delta_mm[X_AXIS] = 0.05 // nice!
delta_mm[Y_AXIS] = 0.1 // also nice!
delta_mm[E_AXIS] = 0.0025 // <> 0.00148..

So our used ratio would be 0.02236, which is 169% of the original value! This means the pressure LIN_ADVANCE advances to is also 169% of the needed one. On the next move, we might round to the oposite value and maybe only get 70% of the ideal ratio, so 30% less pressure than expected.

Of course that's what the extruder will do in real live, the first move will have 169% extrusion compared to what the slicer was telling Marlin and the next on has only 70%. In real world, the spring effect of the filament will safe our print as the pressure will be an average so the delta will be 0 in total. But it's not good at all that the stepper tries to follow this spiky up and down pressure movement at all, we want to keep it calm and pressurize the nozzle to the wanted average pressure.

That's also the reason why we need very high resolution e axis when it comes to fine detail prints like with a 0.2mm nozzle. In this case, the single steps of the extruder are spaced too far away from each other with normal resolution extruders, so the springiness of the filament will not longer safe the print. The result is a highly unstable pulsating flow which ruins the print.

Reply:
So, how about if, when `LIN_ADVANCE` is enabled, we do the `block->millimeters` calculation above using your supplied values instead of the less precise steps-to-mm method? Would that work?

Reply:
```diff
  block->millimeters = SQRT(
    #if CORE_IS_XY
      sq(delta_mm[X_HEAD]) + sq(delta_mm[Y_HEAD]) + sq(delta_mm[Z_AXIS])
    #elif CORE_IS_XZ
      sq(delta_mm[X_HEAD]) + sq(delta_mm[Y_AXIS]) + sq(delta_mm[Z_HEAD])
    #elif CORE_IS_YZ
      sq(delta_mm[X_AXIS]) + sq(delta_mm[Y_HEAD]) + sq(delta_mm[Z_HEAD])
+   #elif ENABLED(LIN_ADVANCE) && !IS_KINEMATIC
+       sq(target_float[X_AXIS] - position_float[X_AXIS])
+     + sq(target_float[Y_AXIS] - position_float[Y_AXIS])
+     + sq(target_float[Z_AXIS] - position_float[Z_AXIS])
    #else
      sq(delta_mm[X_AXIS]) + sq(delta_mm[Y_AXIS]) + sq(delta_mm[Z_AXIS])
    #endif
  );
```

Reply:
This should save some computes:

<details><summary>Replace the `delta_mm` calculation</summary>

```cpp
  #if IS_CORE
    float delta_mm[Z_HEAD + 1];
  #else
    float delta_mm[ABCE];
  #endif
  #if ENABLED(LIN_ADVANCE) && !IS_KINEMATIC
    delta_mm[A_AXIS] = target_float[X_AXIS] - position_float[X_AXIS];
    delta_mm[B_AXIS] = target_float[Y_AXIS] - position_float[Y_AXIS];
    delta_mm[C_AXIS] = target_float[Z_AXIS] - position_float[Z_AXIS];
  #elif CORE_IS_XY
    delta_mm[X_HEAD] = da * steps_to_mm[A_AXIS];
    delta_mm[Y_HEAD] = db * steps_to_mm[B_AXIS];
    delta_mm[Z_AXIS] = dc * steps_to_mm[Z_AXIS];
    delta_mm[A_AXIS] = (da + db) * steps_to_mm[A_AXIS];
    delta_mm[B_AXIS] = CORESIGN(da - db) * steps_to_mm[B_AXIS];
  #elif CORE_IS_XZ
    delta_mm[X_HEAD] = da * steps_to_mm[A_AXIS];
    delta_mm[Y_AXIS] = db * steps_to_mm[Y_AXIS];
    delta_mm[Z_HEAD] = dc * steps_to_mm[C_AXIS];
    delta_mm[A_AXIS] = (da + dc) * steps_to_mm[A_AXIS];
    delta_mm[C_AXIS] = CORESIGN(da - dc) * steps_to_mm[C_AXIS];
  #elif CORE_IS_YZ
    delta_mm[X_AXIS] = da * steps_to_mm[X_AXIS];
    delta_mm[Y_HEAD] = db * steps_to_mm[B_AXIS];
    delta_mm[Z_HEAD] = dc * steps_to_mm[C_AXIS];
    delta_mm[B_AXIS] = (db + dc) * steps_to_mm[B_AXIS];
    delta_mm[C_AXIS] = CORESIGN(db - dc) * steps_to_mm[C_AXIS];
  #else
    delta_mm[A_AXIS] = da * steps_to_mm[A_AXIS];
    delta_mm[B_AXIS] = db * steps_to_mm[B_AXIS];
    delta_mm[C_AXIS] = dc * steps_to_mm[C_AXIS];
  #endif
  delta_mm[E_AXIS] = esteps_float * steps_to_mm[E_AXIS_N];

  if (block->steps[A_AXIS] < MIN_STEPS_PER_SEGMENT && block->steps[B_AXIS] < MIN_STEPS_PER_SEGMENT && block->steps[C_AXIS] < MIN_STEPS_PER_SEGMENT) {
    block->millimeters = FABS(delta_mm[E_AXIS]);
  }
  else if (!millimeters) {
    block->millimeters = SQRT(
      #if !IS_CORE || (ENABLED(LIN_ADVANCE) && !IS_KINEMATIC)
        sq(delta_mm[X_AXIS]) + sq(delta_mm[Y_AXIS]) + sq(delta_mm[Z_AXIS])
      #elif CORE_IS_XY
        sq(delta_mm[X_HEAD]) + sq(delta_mm[Y_HEAD]) + sq(delta_mm[Z_AXIS])
      #elif CORE_IS_XZ
        sq(delta_mm[X_HEAD]) + sq(delta_mm[Y_AXIS]) + sq(delta_mm[Z_HEAD])
      #elif CORE_IS_YZ
        sq(delta_mm[X_AXIS]) + sq(delta_mm[Y_HEAD]) + sq(delta_mm[Z_HEAD])
      #endif
    );
  }
  else
    block->millimeters = millimeters;
```
</details>

Reply:
Ok, so I've added a separate commit that takes the most accurate mm available and applies it right from the start. With `LIN_ADVANCE` we can can use `target_float - position_float` right from the start to get `diff_mm`. Then we get the `SQRT(sq+sq+sq)` distance as usual. This eliminates the need for another SQRT and we can just use…
```cpp
block->e_D_ratio = delta_mm_e / block->millimeters;
```